### PR TITLE
[Program:GCI] Comment task endpoints (Part 2)

### LINKED ISSUE #425: comment a task
# Description:

**Is your feature request related to a problem? Please describe.**
As a User I can not comment task which makes completing task harder.

**Describe the solution you'd like**
As a User I would like to comment a task so I could ask/answer related to task questions.

**Describe alternatives you've considered**
Alternative is using chat.

**Additional context**
We need to refactor current tasks to be able to add comments to tasks.

# PRs:
[398](https://github.com/systers/mentorship-backend/pull/398)
[356](https://github.com/systers/mentorship-backend/pull/356)
[368](https://github.com/systers/mentorship-backend/pull/368)

### PR DESCRIPTION
**Pull request by daksha-aeer**
### Description
Comment endpoints created for:
1. **CREATE**: POST /comment
2. **READ**: GET /comment and /comments
3. **UPDATE**: PUT /comment
4. **DELETE**: DELETE /comment

tasks_list table from Part 1 task used. Comments are validated using relation_id and task_id fields in table.

Fixes #425

### Type of Change:
- Code

**Code/Quality Assurance Only**
- This change requires a documentation update (software upgrade on readme file)
- New feature (non-breaking change which adds functionality pre-approved by mentors)

### How Has This Been Tested?
1. POST /comment
task47_post_1

**Working**
task47_post_2

2. GET 
- GET /comment
task47_get_1

**Working**
task47_get_2

- GET /comments
task47_get_3

**Working**
task47_get_4

3. PUT /comment
task47_put_1

**Working**
task47_put_2

4. Delete
task47_delete_1

**Working**
task47_delete_2

### Checklist:

- [x] My PR follows the style guidelines of this project
- [x] I have performed a self-review of my own code or materials
- [x] I have commented my code or provided relevant documentation, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] Any dependent changes have been merged
- [ ] Update Postman API at /docs folder
- [ ] Update Swagger documentation and the exported file at /docs folder
- [ ] Update requirements.txt

**Code/Quality Assurance Only**
- [ ] My changes generate no new warnings 
- [ ] My PR currently breaks something (fix or feature that would cause existing functionality to not work as expected)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been published in downstream modules

### CODE REVIEW

File: app/database/models/task_comment.py [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+    id = db.Column(db.Integer, primary_key=True, autoincrement=True)
+    request_id = db.Column(db.Integer, nullable=False)
+    task_id = db.Column(db.Integer, nullable=False)
+    user_id = db.Column(db.Integer, nullable=False)
+    creation_time = db.Column(db.Float, nullable=False)
+    modification_time = db.Column(db.Float, nullable=True)
+    comment = db.Column(db.String, nullable=False)
+
+    def __init__(self, request_id, task_id, user_id, comment):
+        """"Initialises TaskCommentModel class with request_id, task_id, user_id, and comment. """
```
Comment (isabelcosta) [inline app/database/models/task_comment.py]:
```suggestion
        """Initialises TaskCommentModel class with request_id, task_id, user_id, and comment. """
```

File: app/api/resources/mentorship_relation.py [hunk 2]
```diff
@@ -335,3 +336,137 @@ def put(cls, request_id, task_id):
...
+        response = CommentDAO.get_task_comment(user_id, request_id, task_id, comment_id)
+
+        return response
+
+    @classmethod
+    @jwt_required
+    @mentorship_relation_ns.doc('delete_comment_for_task')
+    @mentorship_relation_ns.expect(auth_header_parser)
+    @mentorship_relation_ns.response(200,"%s"%messages.COMMENT_WAS_DELETED_SUCCESSFULLY)
+    def delete(cls, request_id, task_id, comment_id):
```
Comment (isabelcosta) [inline app/api/resources/mentorship_relation.py:435]: this api has more possible responses which are not documented here. Can you please add them here?

File: app/api/resources/mentorship_relation.py [hunk 3]
```diff
@@ -335,3 +336,137 @@ def put(cls, request_id, task_id):
...
+        response = CommentDAO.delete_task_comment(user_id, request_id, task_id, comment_id)
+
+        return response
+
+    @classmethod
+    @jwt_required
+    @mentorship_relation_ns.doc('update_comment_for_task')
+    @mentorship_relation_ns.expect(auth_header_parser, create_comment_request_body)
+    @mentorship_relation_ns.response(200,"%s"%messages.UPDATED_COMMENT_WITH_SUCCESS)
+    def put(cls, request_id, task_id, comment_id):
```
Comment (isabelcosta) [inline app/api/resources/mentorship_relation.py:450]: this api has more possible responses which are not documented here. Can you please add them here? Such as 404 for task and comment not found. Can you check for other APIs you created, which might be missing that.

**Review by isabelcosta — CHANGES_REQUESTED** [review]
@daksha-aeer thank you so much for your contribution and following our guidelines.
I requested some changes, can you please address those.

File: app/api/resources/mentorship_relation.py (hunk 3 above)
Comment (daksha-aeer) [reply app/api/resources/mentorship_relation.py:450]: Same as above
image

**Comment by daksha-aeer** [comment]
Changes made, please review : )

File: app/api/resources/mentorship_relation.py [hunk 4]
```diff
@@ -335,3 +336,137 @@ def put(cls, request_id, task_id):
...
+    @mentorship_relation_ns.response(400, '%s'%messages.COMMENT_FIELD_IS_MISSING)
+    @mentorship_relation_ns.response(401, '%s'%messages.USER_NOT_INVOLVED_IN_THIS_MENTOR_RELATION)
+    @mentorship_relation_ns.response(404, '%s\n%s'%(
+        messages.TASK_DOES_NOT_EXIST, messages.MENTORSHIP_RELATION_DOES_NOT_EXIST
+        ))
+    def post(cls, request_id, task_id):
+        """
+        Add a comment for a task.
+        """
+        # TODO check if user id is well parsed, if it is an integer
```
Comment (annabauza) [inline app/api/resources/mentorship_relation.py]: is there anything left TODO?

File: app/api/resources/mentorship_relation.py (hunk 2 above)
Comment (daksha-aeer) [reply app/api/resources/mentorship_relation.py:435]: @isabelcosta these responses(400,401,404) have been documented at the class level, not method level. get(), delete() and put() functions have these responses.
```py
@mentorship_relation_ns.response(400, '%s'%messages.COMMENT_FIELD_IS_MISSING)
@mentorship_relation_ns.response(401, '%s'%messages.USER_NOT_INVOLVED_IN_THIS_MENTOR_RELATION)
@mentorship_relation_ns.response(404, '%s\n%s\n%s'%(
    messages.TASK_DOES_NOT_EXIST, messages.MENTORSHIP_RELATION_DOES_NOT_EXIST,
    messages.COMMENT_DOES_NOT_EXIST
    ))
@mentorship_relation_ns.route('mentorship_relation/<int:request_id>/task/<int:task_id>/comment/<int:comment_id>')
class Comment(Resource):
```

image

File: app/api/resources/mentorship_relation.py (hunk 4 above)
Comment (daksha-aeer) [reply app/api/resources/mentorship_relation.py]: Most endpoints in resources/mentorship_relation.py have this comment. I guess some user_id check will be added in future, so I left the comment in my endpoints as well.

Comment (daksha-aeer) [reply app/api/resources/mentorship_relation.py]: I do not know what it'll actually be used for. Should I remove it or let it be?

**Review by annabauza — CHANGES_REQUESTED** [review]
please have a look at the comment. Thanks

File: app/database/models/task_comment.py (hunk 1 above)
Comment (daksha-aeer) [reply app/database/models/task_comment.py]: Done

**Comment by daksha-aeer** [comment]
> please have a look at the comment. Thanks

Hi Anna I've made the changes. get_jwt_identity() finds out if valid user JWT was used so I removed that comment.

File: app/api/resources/mentorship_relation.py (hunk 4 above)
Comment (daksha-aeer) [reply app/api/resources/mentorship_relation.py]: I removed the comment. The work is done by get_jwt_token() function.

**Review by annabauza — APPROVED** [review]

**Comment by annabauza** [comment]
closing due to #356 merge

**Pull request closed without merging**